src/impls.rs

```diff
@@ -82,21 +83,34 @@ macro_rules! impl_uint_from_fill {
 }
 
 macro_rules! fill_via_chunks {
-    ($src:expr, $dest:expr, $N:expr) => ({
-        let chunk_size_u8 = min($src.len() * $N, $dest.len());
-        let chunk_size = (chunk_size_u8 + $N - 1) / $N;
-
-        // Convert to little-endian:
-        for ref mut x in $src[0..chunk_size].iter_mut() {
-            **x = (*x).to_le();
+    ($src:expr, $dst:expr, $size:expr) => ({
+        let chunk_size_u8 = min($src.len() * $size, $dst.len());
+        let chunk_size = (chunk_size_u8 + $size - 1) / $size;
+        if cfg!(target_endian="little") {
+            // Unsafe code copied from `byteorder::write_slice_native`.
+            unsafe {
+                copy_nonoverlapping(
+                    $src.as_ptr() as *const u8,
+                    $dst.as_mut_ptr(),
+                    chunk_size_u8);
+            }
+        } else {
+            // Unsafe code copied from `byteorder::write_slice` and
+            // `byteorder::write_num_bytes`, with the addition to copy only
+            // `chunk.len()`.
+            // Byteorder assumes we want to copy only complete integers, while
+            // for us the destination may need only a part of the last integer.
+            for (&n, chunk) in $src.iter().zip($dst.chunks_mut($size)) {
+                unsafe {
+                    // N.B. https://github.com/rust-lang/rust/issues/22776
+                    let bytes = transmute::<_, [u8; $size]>(n.to_le());
+                    copy_nonoverlapping((&bytes).as_ptr(),
```

Comment:
`bytes` is an array, so `bytes.as_ptr()` should be the same? I'd prefer the following, though unfortunately it means passing another parameter (or get size from `size_of`):

```
let tmp = n.to_le();
let src_ptr = &tmp as *const $ty as *const [u8; $size];  // inline this is you like
```

I don't think you need to reference that bug; I think Rust's lifetime analysis is good enough now to prevent it.